File: MergeSort/lua/tests/binary_search_test.lua
local binarySearch = require("../src/binary_search")

-- Test helper function
local function assertEqual(expected, actual, message)
    if expected ~= actual then
        error(string.format("%s: expected %s, got %s", message, tostring(expected), tostring(actual)))
    end
end

-- Test cases
local function runTests()
    -- Test with target in middle
    assertEqual(3, binarySearch({1, 2, 3, 4, 5}, 3), "Should find target in middle")

    -- Test with target at start
    assertEqual(1, binarySearch({1, 2, 3, 4, 5}, 1), "Should find target at start")

    -- Test with target at end
    assertEqual(5, binarySearch({1, 2, 3, 4, 5}, 5), "Should find target at end")

    -- Test with target not present
    assertEqual(nil, binarySearch({1, 2, 3, 4, 5}, 6), "Should return nil when target not found")

    -- Test with empty array
    assertEqual(nil, binarySearch({}, 1), "Should handle empty array")

    -- Test with single element
    assertEqual(1, binarySearch({1}, 1), "Should handle single element")

    -- Test with negative numbers
    assertEqual(2, binarySearch({-5, -3, -1, 0, 2}, -3), "Should handle negative numbers")

    -- Test with duplicate elements (should return first occurrence)
    assertEqual(2, binarySearch({1, 2, 2, 2, 3}, 2), "Should find first occurrence")

    print("All tests passed!")
end

-- Run tests
local success, err = pcall(runTests)
if not success then
    print("Test failed: " .. tostring(err))
    os.exit(1)
end

File: MergeSort/cpp/tests/binary_search_test.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include "../src/binary_search.cpp"

// Test framework
#define TEST_ASSERT(condition, message) \
    do { \
        if (!(condition)) { \
            std::cerr << "Test failed: " << message << std::endl; \
            return false; \
        } \
    } while (0)

bool test_binary_search() {
    // Test with integers
    std::vector<int> arr1 = {1, 2, 3, 4, 5};
    TEST_ASSERT(binary_search(arr1, 3) == 2, "Target in middle");
    TEST_ASSERT(binary_search(arr1, 1) == 0, "Target at start");
    TEST_ASSERT(binary_search(arr1, 5) == 4, "Target at end");
    TEST_ASSERT(binary_search(arr1, 6) == -1, "Target not present");

    // Test with empty array
    std::vector<int> empty_arr;
    TEST_ASSERT(binary_search(empty_arr, 1) == -1, "Empty array");

    // Test with single element
    std::vector<int> single_arr = {1};
    TEST_ASSERT(binary_search(single_arr, 1) == 0, "Single element");

    // Test with negative numbers
    std::vector<int> neg_arr = {-5, -3, -1, 0, 2};
    TEST_ASSERT(binary_search(neg_arr, -3) == 1, "Negative numbers");

    // Test with duplicate elements
    std::vector<int> dup_arr = {1, 2, 2, 2, 3};
    TEST_ASSERT(binary_search(dup_arr, 2) == 1, "Duplicate elements");

    // Test with strings
    std::vector<std::string> str_arr = {"apple", "banana", "cherry", "date", "elderberry"};
    TEST_ASSERT(binary_search(str_arr, std::string("cherry")) == 2, "String target in middle");
    TEST_ASSERT(binary_search(str_arr, std::string("apple")) == 0, "String target at start");
    TEST_ASSERT(binary_search(str_arr, std::string("elderberry")) == 4, "String target at end");
    TEST_ASSERT(binary_search(str_arr, std::string("fig")) == -1, "String target not present");

    std::cout << "All tests passed!" << std::endl;
    return true;
}

int main() {
    return test_binary_search() ? 0 : 1;
}
